### DISCUSSION TITLE
Maps in a two-level nested struct do not expand in debug window

### DISCUSSION THREAD

**Discussion by rchapin** [golang/vscode-go#1792] [Help] [answered]
I am using vscode under Debian 10, Xfce 4

Version: 1.60.1
Commit: 83bd43bc519d15e50c4272c6cf5c1479df196a4d
Date: 2021-09-10T17:06:57.931Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 4.19.0-10-amd64

I am unable to see the values in maps in a two-level nested struct.  I can see values in slices/arrays, but not maps.  It seems to be any kind of map.

![debug-double-nested-01](https://user-images.githubusercontent.com/4450872/134175972-d14d6730-7e75-4cab-9737-6c1b0549529a.png)

If there is a reference to the nested struct, I am able to drill down into it to see the map values.

![debug-double-nested-02](https://user-images.githubusercontent.com/4450872/134176068-7dec3d51-d4f8-4d59-ba4e-1c7bf95c0254.png)

Run the following code, placing a breakpoint on the fmt.Println line

```
package main

import "fmt"

type MyMapIface map[string]interface{}
type MyMap map[string]string
type Inner struct {
	Map  map[string]string
	Map2 MyMap
	Map3 MyMapIface
	List []string
}

type Outer struct {
	In Inner
}

func run() {
	/*
		Any maps in a two-level nested struct do NOT expand in the debug window.
	*/
	inner := Inner{
		Map: map[string]string{
			"foo": "blah",
		},
		Map2: MyMap{"this": "that", "the": "other"},
		Map3: MyMapIface{"this": 1, "that": 4, "other": "bar"},
		List: []string{"bar", "baz"},
	}
	outer := Outer{
		In: inner,
	}
	fmt.Printf("outer=%v\n", outer)
}

func main() {
	run()
}
```
Is this a known behavior of vscode?

Thanks,
-- Ryan

**Comment by suzmue** [comment, marked as answer]
This is a known limitation of the legacy debug adapter. To change the levels of variables to recurse you can change [`maxVariableRecurse` in dlvLoadConfig](https://github.com/golang/vscode-go/blob/master/docs/debugging-legacy.md#configuration)

The new debug adapter [automatically loads variables](https://github.com/golang/vscode-go/blob/master/docs/debugging.md#data-inspection) to allow users to drill in to nested variables.  If you are using local debugging, we recommend switching to the new debug adapter using the setting "debugAdapter": "dlv-dap". The `remote attach` request is a work in progress in the new debug adapter and documentation can be found [here](https://github.com/golang/vscode-go/blob/master/docs/debugging.md#remote-debugging).

**Reply by rchapin** [reply]
@suzmue Thank you for getting back to me so quickly.

Changing the ```maxVariableRecurse``` in my settings enables me to drill down into the nested maps without a problem.

```
{
    "workbench.colorTheme": "Default Light+",
    "editor.minimap.enabled": false,
    "editor.fontSize": 12,
    "go.delveConfig": {

        "dlvLoadConfig": {
            "followPointers": true,
            "maxVariableRecurse": 3,
            "maxStringLen": 64,
            "maxArrayValues": 64,
            "maxStructFields": -1
        },
        "apiVersion": 2,
        "showGlobalVariables": true,
        "debugAdapter": "legacy",
        "substitutePath": []
    },
    "workbench.tree.indent": 16,
    "files.autoSave": "afterDelay",
    "go.coverageDecorator": {
        "type": "gutter",
        "uncoveredGutterStyle": "blockred"
    },
    "rewrap.wrappingColumn": 100,
    "editor.renderWhitespace": "all",
    "files.trimTrailingWhitespace": true
}
```
Switching to "dlv-dap" ```debugAdapter``` enables me to also drill-down into the nested maps but without having to specify a max-depth of sort.

```
{
    "workbench.colorTheme": "Default Light+",
    "editor.minimap.enabled": false,
    "editor.fontSize": 12,
    "go.delveConfig": {
        "apiVersion": 2,
        "showGlobalVariables": true,
        "debugAdapter": "dlv-dap",
        "substitutePath": []
    },
    "workbench.tree.indent": 16,
    "files.autoSave": "afterDelay",
    "go.coverageDecorator": {
        "type": "gutter",
        "uncoveredGutterStyle": "blockred"
    },
    "rewrap.wrappingColumn": 100,
    "editor.renderWhitespace": "all",
    "files.trimTrailingWhitespace": true
}
```
However, with ```dlv-dap``` the Local debug variables are not automatically displayed and I have to click on the drop down arrow each time I restart a debugging session.

I looked for the configuration docs for the ```dlv-dap``` adapter but could not find them.  Is there a config for this adapter that will automatically show the Global and Local vars?

Thanks,
-- Ryan

**Reply by polinasok** [reply]
@rchapin 
> However, with `dlv-dap` the Local debug variables are not automatically displayed and I have to click on the drop down arrow each time I restart a debugging session.
> 
> I looked for the configuration docs for the `dlv-dap` adapter but could not find them. Is there a config for this adapter that will automatically show the Global and Local vars?

Unfortunately, only the first scope is expanded the default and there is no way for us to configure that. We can merge or rearrange the scopes. Please see #1783 for the discussion on how this can be improved.

**Comment by hyangah** [comment]
Duplicate of https://github.com/golang/vscode-go/issues/135
Will not create an issue for this.
